libs/execution_context/ebpf_native.c

```diff
@@ -1066,6 +1073,26 @@ ebpf_native_load(
     }
     memcpy(local_service_name, (uint8_t*)service_name, service_name_length);
 
+    local_cleanup.service_name = ebpf_allocate((size_t)service_name_length + 2);
```

Comment:
local_service_name and local_cleanup.service_name seem to be redundant, why do we need both?

Reply:
`local_cleanup.service_name` is a copy which is pre-allocated that is later used in a driver unload workitem, and is freed by the workitem.

`local_service_name` is saved in `ebpf_native_module_t` and is freed when this object is freed.

Reply:
I didn't follow why you need both.  Can object-free and driver-unload happen in either order?   If not, then you only need to store it in the one that happens last if you can find it from the one that happens first.

Reply:
It is possible for module free, and driver unload to happen in either order.

Normally, this is the flow:
1. When all the ebpf programs corresponding to the native module have been unloaded, ebpfcore creates a workitem to unload the driver.
2. Driver unload will cause NMR client detach to happen which will release the "attach" reference (which will be the last reference) on the native module.
3. Once the last ref is released, native module will be freed.

It is possible, however, that the native driver service is being stopped at the same time when ebpfcore is scheduling a work item to unload the driver. This can happen due to:
1. A user manually stopping the native driver service.
2. System shut down

In such a race condition, before the "unload workitem" is executed, it is possible that the native driver gets unloaded, NMR detach happens, and native module is cleaned up.

This is the reason we were earlier creating a copy of the service name to pass to the unload workitem (function `ebpf_native_unload()`). This PR is now pre-allocating that memory to later pass to the workitem.

Reply:
Not sure I follow.   The "unload workitem" can hold its own reference when scheduled, released when executed, which would avoid any double-free and not need a service name copy in the work item, no?

Reply:
native module currently expects n+1 references, where n references are for the n programs loaded from it and 1 reference is the attach reference. This design is needed as we want to explicitly unload / clean up the native module when all its programs have been unloaded (but there is still an attach reference present). This allows the native module to be unloaded when all the programs have been loaded and allows the user mode to reload the same module again. Acquiring an extra reference for workitem will break that logic.

Regarding the 2 service_names, the one present in top level struct (`ebpf_native_module_t`) is only used initially to load the driver. After that it is not used anywhere, except for debugging purposes.
I _can reuse_ it for passing it to workitem (and set it to null in `ebpf_native_module_t` once it is passed to workiem to avoid double free). But the only concern is that sometimes the native modules get stuck (do not get cleaned up) as one of the program references has not been cleaned up. In such a case, having the service name in `ebpf_native_module_t` struct may be helpful in debugging.

I will push a new change to reuse the service_name.

Reply:
Pushed new changes.